<a href="https://colab.research.google.com/github/CPukszta/BI-BE-CS-183-2023/blob/main/HW1/Problem3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Bi/Be/Cs 183 2022-2023: Intro to Computational Biology
TAs: Meichen Fang, Tara Chari, Zitong (Jerry) Wang

**Submit your notebooks by sharing a clickable link with Viewer access. Link must be accessible from submitted assignment document.**

Make sure Runtime $\rightarrow$ Restart and run all works without error

**HW 1 Problem 3**

For this problem you will be exploring the phenomenon known as Simpson's Reversal in a single-cell sequencing dataset.
 
As described in Lecture, Simpson's Reversal describes the phenomenon where a certain trend appears within certain groups of the data, but disappears or is reversed when the groupings are changed. One of the advantages of single-cell RNA sequencing as opposed to bulk RNA sequencing (where all reads from all cells are pooled together) is that it provides the opportunity, in principle, to avoid incorrect inferences about correlation between pairs of genes.

For this problem you will mine a single-cell RNA-seq dataset to see if this phenomenon occurs in practice.

In [2]:
#To run a code cell, select the cell and hit Command/Ctrl+Enter or click the run/play symbol
#Click Insert --> Code Cell or the '+ Code' option to insert a new code cell

In [3]:
#Click Insert --> Text Cell or the '+ Text' option to insert a cell for text as below

In [4]:
# This is  used to time the running of the notebook
import time
start_time = time.time()

Text here for descriptions, explanations, etc

##**Import data and install packages**

In [5]:
import numpy as np
import scipy.io as sio
import pandas as pd
import bokeh.io
import bokeh.plotting
import matplotlib.pyplot as plt #Can use other plotting packages like seaborn


bokeh.io.output_notebook()


In [6]:
#Download count matrix of cell by gene counts and metadata files, DOI: 10.22002/D1.20031

import requests
from tqdm import tnrange, tqdm_notebook
def download_file(doi,ext):
    url = 'https://api.datacite.org/dois/'+doi+'/media'
    r = requests.get(url).json()
    netcdf_url = r['data'][0]['attributes']['url']
    r = requests.get(netcdf_url,stream=True)
    #Set file name
    fname = doi.split('/')[-1]+ext
    #Download file with progress bar
    if r.status_code == 403:
        print("File Unavailable")
    if 'content-length' not in r.headers:
        print("Did not get file")
    else:
        with open(fname, 'wb') as f:
            total_length = int(r.headers.get('content-length'))
            pbar = tnrange(int(total_length/1024), unit="B")
            for chunk in r.iter_content(chunk_size=1024):
                if chunk:
                    pbar.update()
                    f.write(chunk)
        return fname

download_file('10.22002/D1.20031','.tar.gz')

!tar -xvf D1.20031.tar.gz

<ipython-input-6-07df2730af8c>:20: TqdmDeprecationWarning: Please use `tqdm.notebook.trange` instead of `tqdm.tnrange`
  pbar = tnrange(int(total_length/1024), unit="B")


  0%|          | 0/6617 [00:00<?, ?B/s]

._cell_metadata.csv
cell_metadata.csv
._gene_metadata.csv
gene_metadata.csv
._exutero.mtx
exutero.mtx


## **Read in data for analysis**

**The dataset**

Much of mammalian embryo development occurs after the embryo implants in the uterus, however this makes the embryo relatively inaccessible and thus hard to study. In [Aguilera-Castrejon et al., 2021](https://www.nature.com/articles/s41586-021-03416-3#Sec7), a cell culture platform was established to culture post-implantation embryos ex utero (outside the uterus) from pre-gastrulation until himdlimb formation (organogenesis).  Single-cell sequencing was done for both in utero and ex utero cultured mouse embryos, to compare development of the externally cultured cells.

<center><img src="https://drive.google.com/uc?export=view&id=1b6BM_W5Nk2r31RdGE7-bKDuObQq279qh" alt="EMFigure" width="800" height="350"><center>


**The count matrix**

This matrix is 6,205 cells by 2,000 genes. We are only working with the ex-utero cultured mouse embryo cells here.

1.   For each cell, gene counts were normalized to have the same number of total counts (usually 1e5 or 1e6), with cell-gene counts thus scaled accordingly.

2.   Counts were then log-normalized, using the log(1+x), where x is each cell's gene count. The 1 accounts for 0 count genes. 

3. The top ~2000 genes were selected for those that displayed large variance in expression amongst the cells ('highly variable genes').



In [7]:
#Get gene count matrix
count_mat = sio.mmread('exutero.mtx')

count_mat.shape

(6205, 2000)

In [8]:
#Get metadata dataframe for the 6,205 cells (rows of the matrix)
meta = pd.read_csv('cell_metadata.csv',index_col = 0)
print(pd.unique(meta["cell_type"]))

meta['cell_type'].value_counts()

['Mixed Mesoderm' 'Blood' 'Neural Tube' 'Pharyngeal Mesoderm'
 'Extra-Embryonic Ectoderm' 'Endothelial' 'Extra-Embryonic Endoderm'
 'Amnion' 'Presomitic Mesoderm' 'Cardiac' 'Mid Hind Brain' 'Placodes'
 'Somitic Mesoderm' 'Foregut' 'Neural Crest' 'Mid Hind Gut'
 'Extra-Embryonic Mesoderm']


Blood                       1104
Mixed Mesoderm               551
Endothelial                  547
Foregut                      510
Amnion                       443
Mid Hind Brain               429
Extra-Embryonic Endoderm     406
Placodes                     372
Presomitic Mesoderm          366
Pharyngeal Mesoderm          319
Cardiac                      317
Extra-Embryonic Ectoderm     284
Neural Tube                  216
Somitic Mesoderm             172
Neural Crest                  80
Extra-Embryonic Mesoderm      51
Mid Hind Gut                  38
Name: cell_type, dtype: int64

In [9]:
#Get metadata dataframe for the 2,000 genes (columns of the matrix)

meta_gene = pd.read_csv('gene_metadata.csv',index_col = 0)
meta_gene.head()


gene_name
0     Sox17
1    Lypla1
2      Sgk3
3      Xkr9
4      Bag2

## **Problem 3 (36 points)** 
Find an example of Simpson's reversal.

Generally Simpson's reversal describes how a trend may appear in several groups of data, but not exist or be reversed when groups are combined.

Here we will focus on trends of correlation, where two variables (genes) appear correlated in one direction (e.g. **positively**) given one set of groupings, but display **no or negative** correlation in another set of groupings.

For the purposes of this homework, we will be using the common Pearson correlation coefficient to measure correlation between variables.

This can be calculated using numpy's corrcoef() function in python, where correlation values are between -1 and 1.

In [10]:
#Use np.corrcoef()

#Given two lists/vectors of values x and y, np.corrcoef(x,y) will return a 
#pairwise correlation matrix where the diagonal represent correlation between
#x with x or y with y, (which are 1) and the off-diagonals which both represent  
#correlation of x with y. 

#The off-diagonal value is the value we are interested in

x = [1.0,2.0,3.0,4.0]
y = [4.0,5.0,2.0,6.0]

corrs = np.corrcoef(x,y) 
print('Correlation matrix: \n', corrs)

print('Corr in first row, second column: ', corrs[0,1])

Correlation matrix: 
 [[1.         0.22677868]
 [0.22677868 1.        ]]
Corr in first row, second column:  0.22677868380553634


**Problem:** Find two genes which display a Simpson's reversal in their correlations depending on how they are grouped. We are calculating correlation between the genes' expression vectors i.e. their values across cells.

Examples of grouping comparison include, but are not limited to:

*   Correlation of two genes within each one of two cell types versus their correlations within the two cell types (no cell type grouping)
*   Correlation of two genes within one cell type versus their correlation across all cells (no cell type grouping)


*Try to report the most extreme difference in correlation you can find*

Show:

a.   All code for subsetting the count matrix and calculating correlation values for a pair of genes. (12 points)

Code can be added to code cells below. Feel free to add more code cells if necessary.






In [11]:
# ------- Example code for finding gene-pair correlations within a cell type versus across all cells -----------

gene1 = 'Krt8' # Keratin 8 is a member of the type II keratin family
gene2 = 'Hmgn5'  #High Mobility Group Nucleosome Binding Domain 5

indices = meta['cell_type'].isin(['Mid Hind Gut']) #Get indices for where the cell_type is Blood only

g1_indices = meta_gene['gene_name'].tolist().index(gene1) #Get exact index for where gene names are gene1 only
g2_indices = meta_gene['gene_name'].tolist().index(gene2)

#Correlation of counts within cell type
count_sub_g1 = count_mat[indices,g1_indices] #Filter rows (cells) of matrix
count_sub_g2 = count_mat[indices,g2_indices]

corrs = np.corrcoef(count_sub_g1,count_sub_g2)
print('Correlation within type: ', corrs[0,1])

#Correlation of counts within cell type
count_all_g1 = count_mat[:,g1_indices] #Filter rows (cells) of matrix Sntg1
count_all_g2 = count_mat[:,g2_indices]

corrs = np.corrcoef(count_all_g1,count_all_g2)
print('Correlation across all cells: ',corrs[0,1])

#We see positive correlation within the Mid Hind Gut cell type but small correlation across all cells

print(corrs)

Correlation within type:  0.43329374882242444
Correlation across all cells:  0.07308318924730003
[[1.         0.07308319]
 [0.07308319 1.        ]]


I'm most interested in the mesoderm and the subsequent differentiations that occur after mesoderm induction. Therefore, lets look for a reversal in the mesoderm populations! I will start by calculating the correlations for all mesoderm cells. I am only running this on the first 200 genes because this is computaionally expensive and does not scale very easily. 

In [148]:
# ------- Your code below ------- 
n=200
cor_all = np.zeros((n,n))
indicies_pooled = meta['cell_type'].isin(['Presomitic Mesoderm','Pharyngeal Mesoderm','Somitic Mesoderm','Mixed Mesoderm',"Extra-Embryonic Mesoderm"]) 

for i in range(n):
  for j in range(n):
    g1 = count_mat[indicies_pooled,i]
    g2 = count_mat[indicies_pooled,j]
    corrs = np.corrcoef(g1,g2)
    cor_all[i,j] = corrs[0,1]


Now lets calcualate the correlations for the presomitic mesoderm. 

In [149]:
# now lets find this same matrix for just one cell type
#lets choose Presomitic Mesoderm
cor_blood = np.zeros((n,n))
indices = meta['cell_type'].isin(['Presomitic Mesoderm'])

for i in range(n):
  for j in range(n):
    g1 = count_mat[indices,i]
    g2 = count_mat[indices,j]
    corrs = np.corrcoef(g1,g2)
    cor_blood[i,j] = corrs[0,1]

Finally we can calcualte the difference between the two correlations matricies and choose the largest differences. This will give us a starting point of genes. I will set a threshold for the differences which will give some genes to try and graph.

In [183]:
# now we can find the difference between the two correlation values for each gene
#from there we will choose the two largest differences and plot the gene expression values

diff = np.zeros((n,n))
diff = np.abs(cor_blood-cor_all)
x,y = np.where(diff > 0.60)
print(x)
print(y)

[ 26  44  71  99 129 160]
[ 71 160  26 129  99  44]


Now we can choose the genes of interest and grab their values for plotting!

In [178]:
# lets plot our reversals
#start by grabbing the gene names for the two selected genes
#After some examination I will choose the combination of gene 99 and 129 above
gene1 = meta_gene["gene_name"][99 ]
gene2 = meta_gene["gene_name"][129]

print(gene1,gene2)

g1_indices = meta_gene['gene_name'].tolist().index(gene1) 
g2_indices = meta_gene['gene_name'].tolist().index(gene2)

count_sub_g1_all = count_mat[indicies_pooled,g1_indices]
count_sub_g2_all = count_mat[indicies_pooled,g2_indices]


#now grab the values for blood cells
count_sub_g1_blood = count_mat[indices,g1_indices]
count_sub_g2_blood = count_mat[indices,g2_indices]



Apoa2 Cubn


Lets plot!

In [191]:
p = bokeh.plotting.figure(
    height=500,
    width=700,
    x_axis_label=gene1,
    y_axis_label=gene2,
    title = "transcript counts"
)

p.circle(count_sub_g1_all,count_sub_g2_all,color="hotpink",legend_label="Mesoderm")
p.circle(count_sub_g1_blood,count_sub_g2_blood,color="darkturquoise",legend_label="Presomitic Mesoderm")

p.legend.click_policy="hide"

bokeh.io.show(p)

This plot looks pretty nice. There are certainly less cells of the presomatic mesoderm that express both genes. It is therfore, easy to see how the presomatic mesoderm cells would have a lower correlation than the overall mesoderm population. 

Finally, lets print our correlations.

In [192]:
print("The Correlation between", gene1, "and", gene2, "in all Mesoderm is", cor_all[99,129])
print("The Correlation between", gene1, "and", gene2, "in Pharyngeal Mesoderm cells is", cor_blood[99,129])

The Correlation between Apoa2 and Cubn in all Mesoderm is 0.7228096107136631
The Correlation between Apoa2 and Cubn in Pharyngeal Mesoderm cells is 0.09721744142153771


Report: 

b.   What the comparison groupings are (e.g. within one cell type versus across all cells) (6 points)

**My two comparison groups are all mesoderm cells and Pharyngeal Mesoderm cells**

c.   Which two genes you are looking at (6 points)

**I decided to look at Apoa2 and Cubn.**

d.   What the gene-gene correlations are in each grouping (6 points)

**The Correlation between Apoa2 and Cubn in all Mesoderm is 0.7228096107136631**

**The Correlation between Apoa2 and Cubn in Pharyngeal Mesoderm cells is 0.09721744142153771**

e.   Qualitatively comment on how significant or relevant you think these correlation differences are, given the groupings you chose and the correlation values. (6 points)

The reversal seems pretty distinct given that 0.09 within the range of no correlation for the Pharyngeal Mesoderm cells whereas the overal population of mesoderm cells have a fairly strong correlation of 0.72. This might indicate that Pharyngeal Mesoderm does not strongly correlate the expression of Apoa2 and Cubn, but other mesoderm-derived populations do. 

*Answers b-e can be added directly to this text box*


In [193]:
# Running time of the notebook
print("{:.2f} minutes".format((time.time()-start_time)/60))

220.90 minutes
